announcements: integration with notifications #4378

lavanya-sainik-ericsson opened the pull request:
## Hey, I just made a Pull Request!

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#creating-changesets))
- [X] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/community-plugins/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

lavanya-sainik-ericsson commented:
Hello @kurtaking  / All,

I have to raise this new PR for issue https://github.com/backstage/community-plugins/issues/4378 . as  the old PR https://github.com/backstage/community-plugins/pull/4793 had merge conflicts and it was not allowing me to resolve them locally so I have to use the UI option "Discard 1 commit" & then create this new PR again. So I have all your comments included. Kindly have a look at this PR and let me know if any comments / concerns

04kash commented:
Looks like you need to run `yarn install` and commit your `yarn.lock` file to fix the failing CI

lavanya-sainik-ericsson commented:
Hello @04kash and All, 

Thanks for your comment. I have added the yarn.lock file . 

However, when I tried to amend yarn.lock with my original commit of 7 files, in my local branch after syncing the fork, 3 new  commits were there and now all those  files are also added with my original 8 files changes along with new reviewers list. 
Let me know if this is an expected behavior or if these files need to be removed ? 

If yes, then kindly guide me how to remove /  proceed further .

kurtaking commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -38,6 +38,44 @@ import {
+  notifications.send({
+    recipients: {
+      type: 'entity',
+      entityRef: entityRefsTargets ?? [],
```
@gaelgoth - any thoughts on this?

gaelgoth replied:
Thanks for the changes. Fine to leave it as `broadcast` by default for now 👍🏾. We can introduce recipient selection from the frontend in a separate PR

gaelgoth commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -38,6 +38,44 @@ import {
+    payload: {
+      title: 'New Announcement',
+      description: 'New announcement has been added',
+      link: hostUrl
```
Is there a use case for having a different host? I think setting the path to  `/announcements/view/${announcementId}`  should be sufficient

lavanya-sainik-ericsson replied:
Done

gaelgoth commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -38,6 +38,44 @@ import {
 import { signalAnnouncement } from './service/signal';
 import { AnnouncementsContext } from './service';
+import { NotificationService } from '@backstage/plugin-notifications-node';
+import { Config } from '@backstage/config';
```
Following my comment about the `hostUrl` and the ones regarding the recipient, we can avoid using the config

lavanya-sainik-ericsson replied:
Done

gaelgoth reviewed (commented):
It would be great if you could [set up Notifications ](https://backstage.io/docs/notifications/#installation) on the Backstage dev instance 🙏🏾. This will help with trying/testing notifications locally.

lavanya-sainik-ericsson commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -38,6 +38,44 @@ import {
+      'announcements.notification',
+    );
+
+  const notifications = announcementNotificationOpts.notifications;
```
not needed now based on @gaelgoth next comment. But do revert if nay concern.

lavanya-sainik-ericsson commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -38,6 +38,44 @@ import {
+import { NotificationService } from '@backstage/plugin-notifications-node';
+import { Config } from '@backstage/config';
+
+const sendAnnouncementNotification = (
```
Done

lavanya-sainik-ericsson commented:
Hello @kurtaking , @gaelgoth 

Thanks a lot for your comments. Have updated the PR with the comments included. Kindly review and do let me know if any further concerns or improvements.

Also @gaelgoth , could you please guide on how to set up the Backstage dev instance. I am happy to do that but don't know the instructions and link of instance. So kindly guide here.

kurtaking commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -209,6 +211,9 @@ export async function createRouter(
         });
 
+      // Send announcement notification
+      sendAnnouncementNotification(req, announcement.id, notifications);
```
Also, what happens if we call both `signalAnnouncement` and `sendAnnouncementNotification`. Does that cause two separate signals to go out? 

I'm wondering if they have notifications enabled, we call `sendAnnouncementNotification`, else if they have have signals enabled, we call `signalAnnouncement`, else do nothing.

lavanya-sainik-ericsson replied:
Have updated the code based on the suggestion. Kindly have a look .

gaelgoth commented:
> Hello @kurtaking , @gaelgoth
> 
> Thanks a lot for your comments. Have updated the PR with the comments included. Kindly review and do let me know if any further concerns or improvements.
> 
> Also @gaelgoth , could you please guide on how to set up the Backstage dev instance for you to test as you have mentioned in the comment ? I am happy to do that but don't know link of instance and if you have any specific instructions as we have tested in our local backstage installation. So kindly guide here .

Hi @lavanya-sainik-ericsson, 

You can follow the installation guide here: https://backstage.io/docs/notifications/#installation

Run the `yarn add` commands from the announcements root workspace `cd workspaces/announcements`, and then add the notifications [backend](https://backstage.io/docs/notifications/#add-notifications-backend) and [frontend](https://backstage.io/docs/notifications/#add-notifications-backend) as described in the documentation.

The signals backend is already installed, you can skip that step.

I did try the set up on my side, so far it's look good 😄 

<img width="1263" height="878" alt="image" src="https://github.com/user-attachments/assets/9f51afbb-bba9-4493-a1d6-a37b991f7c01" />

gaelgoth reviewed (commented):
I think it would be necessary to update the [documentation](https://github.com/backstage/community-plugins/blob/main/workspaces/announcements/plugins/announcements-backend/README.md#installation) to mention the prerequisites for Notifications

lavanya-sainik-ericsson commented:
Hello @kurtaking @gaelgoth ,

Thanks for your comments and guidance. Have updated the files. Kindly have and look and let me know if any further comments / concerns.

kurtaking commented on workspaces/announcements/plugins/announcements-backend/src/router.ts:
```diff
@@ -218,9 +228,13 @@ export async function createRouter(
         });
 
-        await signalAnnouncement(announcement, signals);
+        if (isSignalsEnabled) {
+          await signalAnnouncement(announcement, signals);
+        }
+        if (isAnnouncementNotificationsEnabled) {
+          sendAnnouncementNotification(announcement, notifications);
+        }
```
Hey @drodil - do you have time to take a look at this? I'm worried we are going to duplicate signals. 

Should we do a conditional if/elseif here? Meaning if they have announcements enabled, we should only call `sendAnnouncementNotification` and not `signalAnnouncement`. If announcements are disabled, then handle signals like we do currently.

drodil replied:
Hi. It depends. If both are enabled, the UI will show both the announcement and the notification in real time (banner on top of the screen for announcement and pop-up in the lower right corner for notification). 

At the moment, there's no possibility to skip the signal sending in the notifications plugin which IMO would be the right way so that people who are not in the portal at the time of the sending would get a notification about the announcement and see it once they log in. 

Without the possibility to skip the signal sending in notifications, I would send both in this case (of course based on config), add support for this feature in the backstage, and later use it here for notifications. WDYT?

kurtaking replied:
Leaning towards a simple config option to enable/disable sending a notification being sufficient. If you enable notifications on the announcements plugins, it can be assumed you will get both. 

I appreciate you taking the time to review.

kurtaking commented on workspaces/announcements/plugins/announcements-backend/README.md:
```diff
@@ -103,3 +103,31 @@ const response = await fetch(
+Entity: sent to specific users or groups (e.g., owners of an entity).
+
+The Notification plugin delivers real-time alerting, with backend/frontend components to send and display notifications - including push signals.
+To trigger alerts when a new announcement appears, you can combine these systems: send a notification via the Notifications backend whenver an announcement is created.
```
```suggestion
To trigger alerts when a new announcement appears, you can combine these systems: send a notification via the Notifications backend whenever an announcement is created.
```

kurtaking commented on workspaces/announcements/plugins/announcements-backend/README.md:
Also add a bullet point [here](https://github.com/backstage/community-plugins/blob/1def230bd51d25fa1e602729f3af6cb9833eed13/workspaces/announcements/plugins/announcements-backend/README.md?plain=1#L3-L9) for notification support.

kurtaking commented on workspaces/announcements/plugins/announcements-backend/README.md:
```diff
@@ -103,3 +103,31 @@ const response = await fetch(
+
+### Notifications for Announcements
+
+Backstage’s Notification System empowers plugins and services to deliver alerts to users and is directly visible in the UI or via external channels. It supports two notification types:
+Broadcast: sent to all users.
+Entity: sent to specific users or groups (e.g., owners of an entity).
```
Thinking we should point to the notifications docs for more information to keep this specific to announcements. 

```suggestion
Backstage’s Notification System empowers plugins and services to deliver alerts to users and is directly visible in the UI or via external channels. Visit the [docs](https://backstage.io/docs/notifications/) on notifications for more information. 
```

lavanya-sainik-ericsson replied:
Done

kurtaking commented on workspaces/announcements/plugins/announcements-backend/README.md:
````diff
@@ -103,3 +103,31 @@ const response = await fetch(
+The Notification plugin delivers real-time alerting, with backend/frontend components to send and display notifications - including push signals.
+To trigger alerts when a new announcement appears, you can combine these systems: send a notification via the Notifications backend whenver an announcement is created.
+
+Signals and Announcement notifications are disabled by-default (enabled: false).
+It can be enabled in app-config.yaml file when required.
+
+To enable Signals, need to use below configuration under backend:
+
+```yaml
+backend:
+  signals:
+    enabled: true
+```
````
Let's keep this focused on announcements. Enabling signals for announcements [is optional](https://backstage.io/docs/notifications/#optional-add-signals).

lavanya-sainik-ericsson replied:
Done

kurtaking reviewed (commented):
Another round of review. Thank you for working through this with us.

lavanya-sainik-ericsson commented:
Hello @kurtaking ,

. Thanks for your comment. Have updated the documentation as per your suggestion.  Also have left comments on ongoing discussion under https://github.com/backstage/community-plugins/issues/4378#issuecomment-3173459041. Kindly have a look and let me  know if any further comments or concerns.

lavanya-sainik-ericsson commented:
Hello @kurtaking , 

Thanks a lot for your guidance on the issue. We will re analyze the changes based on the given suggestion and will get back accordingly.

lavanya-sainik-ericsson commented:
Hello @kurtaking , @drodil , @gaelgoth ,

We have shared the changes based on send Notification option. Build is failing because of tsc errors. Sorry for that. We will fix it and repush. In the meantime if you have anymore comments, please share.  Have kept signal implementation as it is but if something more is needed, kindly guide as we are unable to test signaling.

lavanya-sainik-ericsson commented:
Hello @kurtaking , @drodil , @gaelgoth ,

We have shared the changes based on send Notification option. No build errors. 
Kindly review it based on your availability and let us know in case of any further comments / concerns.

gaelgoth commented on workspaces/announcements/plugins/announcements-backend/src/service/announcementNotification.ts:
```diff
@@ -0,0 +1,37 @@
+import { NotificationService } from '@backstage/plugin-notifications-node';
+import { AnnouncementModel } from './model';
+
+export const sendAnnouncementNotification = (
+  announcement: AnnouncementModel,
+  notifications?: NotificationService,
+) => {
+  if (!notifications) {
+    return;
+  }
+
+  notifications.send({
+    recipients: {
+      type: 'broadcast',
+    },
+    payload: {
+      title: `New Announcement "${announcement.title}"`,
+      description: announcement.excerpt,
+      link: `/announcements/view/${announcement.id}`,
+    },
+  });
+};
```
To keep consistency with `signals.ts`. I propose to await `notifications.send()` and make it `async` instead of fire-and-forget
```suggestion
export const sendAnnouncementNotification = async (
  announcement: AnnouncementModel,
  notifications?: NotificationService,
) => {
  if (!notifications) {
    return;
  }

   await notifications.send({
    recipients: {
      type: 'broadcast',
    },
    payload: {
      title: `New Announcement "${announcement.title}"`,
      description: announcement.excerpt,
      link: `/announcements/view/${announcement.id}`,
    },
  });
};
```

lavanya-sainik-ericsson replied:
Done

gaelgoth commented on workspaces/announcements/plugins/announcements-common/src/types.ts:
```diff
@@ -102,6 +104,8 @@ export type AnnouncementsFilters = {
   sortBy?: 'created_at' | 'start_at';
   /** Sorting order: "asc" for ascending or "desc" for descending */
   order?: 'asc' | 'desc';
+  /** Whether the notificaion is enabled */
```
```suggestion
  /** Whether the notification is enabled */
```

lavanya-sainik-ericsson replied:
Done

gaelgoth reviewed (changes requested):
Hi @lavanya-sainik-ericsson, thank you for adding the changes we discussed in https://github.com/backstage/community-plugins/issues/4378. I’ve added a few comments. Looks like there are still some unresolved comment with the documentation.

Also, I noticed that some commits not authored by you were modified in your branch. This likely happened due to a `git commit --amend` or a `git rebase` on commits that weren’t yours 🤔

Could you please fix the git history

kurtaking commented on workspaces/announcements/plugins/announcements-common/report.api.md:
```diff
@@ -43,6 +44,7 @@ export type AnnouncementsFilters = {
   active?: boolean;
   sortBy?: 'created_at' | 'start_at';
   order?: 'asc' | 'desc';
+  sendNotification?: boolean;
```
There is probably a better place for this but that can be for another time. 

Just sort of feels off it's on a filter type.

lavanya-sainik-ericsson replied:
Thanks @kurtaking for approving. No don't think so any follow up items left, have tried to include all the comments and suggested changes . But please feel free to share if any further comments or concerns.

kurtaking reviewed (approved):
Hey @lavanya-sainik-ericsson, this is looking good! I think we are ready to merge what you have and iterate from there. 🎉 

Any known follow up items?

lavanya-sainik-ericsson commented:
Hello @gaelgoth ,

Thanks for your comment and highlighting this.  The problem seems to be  with 4 commits. Tried fixing this using interactive rebase but getting errors. We have tried few other hacks but unable to edit these commit messages thus cannot remove this lines.

`Sample error :-

The previous cherry-pick is now empty, possibly due to conflict resolution.
If you wish to commit it anyway, use:

    git commit --allow-empty

Otherwise, please use 'git rebase --skip'
interactive rebase in progress; onto 26b870405
Last commands done (2 commands done):
   pick ae52af241 announcements: integration with notifications #4378
   reword 38e1aaff8 nexus-repository-manager: remove product theme from dev dependencies and dev app (#4798)
`
So is it ok if I create a fresh PR and repush our agreed changes and reference it here or do let me know if you have any more suggestions?
